dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java

```diff
@@ -174,18 +182,147 @@ public void verifyWithProgramStageAndNumericDataElementAndFilter2() {
+            + "LEFT JOIN relationshiptype rty on rty.relationshiptypeid = r.relationshiptypeid WHERE rty.relationshiptypeid = "
+            + relationshipTypeA.getId() + ")) )) " + "as \"" + programIndicatorA.getUid() + "\"  "
+            + "from analytics_enrollment_" + programA.getUid()
+            + " as ax where enrollmentdate >= '2015-01-01' and enrollmentdate <= '2017-04-08' and (uidlevel0 = 'ouabcdefghA' ) limit 101";
+
+        assertSql( sql.getValue(), expected );
+    }
+
+    @Override
+    String getTableName() {
         return this.TABLE_NAME;
     }
 
+    private RelationshipType createRelationshipType(RelationshipEntity fromConstraint,
+                                                    RelationshipEntity toConstraint) {
+        RelationshipType relationshipTypeA = new BeanRandomizer().randomObject(RelationshipType.class);
+
+        RelationshipConstraint from = new RelationshipConstraint();
+        from.setRelationshipEntity(fromConstraint);
+
+        RelationshipConstraint to = new RelationshipConstraint();
+        to.setRelationshipEntity(toConstraint);
+
+        relationshipTypeA.setFromConstraint(from);
+        relationshipTypeA.setToConstraint(to);
+        return relationshipTypeA;
+    }
+
+    private RelationshipType createRelationshipType( RelationshipEntity fromToConstraint )
+    {
+        return createRelationshipType( fromToConstraint, fromToConstraint );
+    }
+
     private void assertSql( String actual, String expected )
     {
-        assertThat( "select " + DEFAULT_COLUMNS + "," + actual, is( expected ) );
+        assertThat( actual, is("select " + DEFAULT_COLUMNS + "," + expected ) );
+    }
+
+    @Test
+    public void verifyWithProgramIndicatorAndRelationshipTypeBothSidesTei2()
+    {
+        Date startDate = getDate( 2015, 1, 1 );
+        Date endDate = getDate( 2017, 4, 8 );
+        Program programB = createProgram( 'B' );
+        String piSubquery = "distinct psi";
+
+        ProgramIndicator programIndicatorA = createProgramIndicator( 'A', programB, "", "" );
+
+        RelationshipType relationshipTypeA = createRelationshipType( RelationshipEntity.TRACKED_ENTITY_INSTANCE );
+
+        EventQueryParams.Builder params = new EventQueryParams.Builder(
+            createRequestParams( programIndicatorA, relationshipTypeA ) ).withStartDate( startDate )
+                .withEndDate( endDate );
+
+        when( programIndicatorService.getAnalyticsSql( "", programIndicatorA, getDate( 2000, 1, 1 ),
+            getDate( 2017, 4, 8 ), "subax" ) ).thenReturn( piSubquery );
+
+        subject.getEnrollments( params.build(), new ListGrid(), 100 );
+
+        verify( jdbcTemplate ).queryForRowSet( sql.capture() );
+
+        String expected = "ax.\"monthly\",ax.\"ou\",(SELECT avg (" + piSubquery + ") FROM analytics_event_"
+            + programB.getUid().toLowerCase() + " as subax WHERE  "
+            + "ax.tei in (select tei.uid from trackedentityinstance tei "
```

Comment:
Was here expecting the join to link ax.tei to the correct subax row through the relationship.
Link to cell in K-map: https://docs.google.com/spreadsheets/d/18u4ZPB_11CfAlxtEz3G85nwmI82hgncHHgQcmVL_B6w/edit#gid=0&range=D3